cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -536,6 +554,304 @@ aggregate_reader_metadata::aggregate_reader_metadata(
+  }
+
+  return schema;
+}
+
+void aggregate_reader_metadata::apply_arrow_schema()
+{
+  // Collect the arrow schema from the key value section of Parquet metadata
+  auto arrow_schema_root = collect_arrow_schema();
+
+  // Check if empty arrow schema collected
+  if (arrow_schema_root.type.id() == type_id::EMPTY and arrow_schema_root.children.size() == 0) {
+    return;
+  }
+
+  // Function to verify equal num_children at each level in Parquet and arrow schemas.
+  std::function<bool(arrow_schema_data_types&, int)> validate_schemas =
+    [&](arrow_schema_data_types& arrow_schema, int schema_idx) {
+      auto& schema_elem = per_file_metadata[0].schema[schema_idx];
+
+      // ensure equal number of children first to avoid any segfaults in children
+      if (schema_elem.num_children == static_cast<int32_t>(arrow_schema.children.size())) {
+        // true if and only if true for all children as well
+        return std::all_of(thrust::make_counting_iterator(0),
+                           thrust::make_counting_iterator(schema_elem.num_children),
+                           [&](auto const& idx) {
+                             return validate_schemas(arrow_schema.children[idx],
+                                                     schema_elem.children_idx[idx]);
+                           });
+      } else {
+        return false;
+      }
+    };
+
+  // Function to co-walk arrow and parquet schemas
+  std::function<void(arrow_schema_data_types&, int)> co_walk_schemas =
+    [&](arrow_schema_data_types& arrow_schema, int schema_idx) {
+      auto& pq_schema_elem = per_file_metadata[0].schema[schema_idx];
+      std::for_each(thrust::make_counting_iterator(0),
+                    thrust::make_counting_iterator(pq_schema_elem.num_children),
+                    [&](auto const& idx) {
+                      co_walk_schemas(arrow_schema.children[idx], pq_schema_elem.children_idx[idx]);
+                    });
+
+      // true for DurationType columns only for now.
+      if (arrow_schema.type.id() != type_id::EMPTY) {
+        pq_schema_elem.arrow_type = arrow_schema.type.id();
+      }
+    };
+
+  // Get Parquet schema root
+  auto pq_schema_root = get_schema(0);
+
+  // verify equal number of children for both schemas at root level
+  if (pq_schema_root.num_children != static_cast<int32_t>(arrow_schema_root.children.size())) {
+    CUDF_LOG_ERROR("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // Verify equal number of children at all sub-levels
+  if (not std::all_of(thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(pq_schema_root.num_children),
+                      [&](auto const& idx) {
+                        return validate_schemas(arrow_schema_root.children[idx],
+                                                pq_schema_root.children_idx[idx]);
+                      })) {
```

Comment:
looks like we have the same pattern in a few places - make_counting_iterator that indexes into arrow_schema_root.children and arrow_schema_root.children
we could declare a zip iterator once and replace all these counting iterators; then we can maybe more easily justify not using raw for loops instead of `for_each` calls :D

Reply:
All done!